### PR TITLE
f4 spi support

### PR DESCRIPTION
**Pull request by lysseslukker**

### CODE REVIEW

**Comment by alfredch** [comment]
Let me know in case I can help <ou with testing and debugging. I did that kind of work recently.

**Comment by lysseslukker** [comment]
Alfredch, what is you approach to the testing? I only have a stm32f411re right now

**Comment by alfredch** [comment]
Testing on two levels is needed: 
1) In the project drone-stm32-map, in file test/periph-macros.rs, add the targets with the feature 'spi' and run `just test-all`. Before that is done, you can set one of the targets in question in file 'Justfile', for instance
stm32_mcu := 'stm32f411' and run `just test`. This will reveal any syntax errors or mismatches with the SVD in your f4.rs macro declarations.
2) Write a basic spi device driver and a sample firmware, for instance, one that echoes generated characters (needs master and slave mode) or talk to an external device which is either slave or master.
I haven't checked the differences of the spi peripheral across the various members of the stm32f4 series, but maybe it's ok to run the `test-all` and verify a driver on the board that we have available (the stm32f401 and stm32f411 on my side).

File: src/periph/spi/f4.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    APB1LPENR,
+    SPI2EN,
+    SPI2RST,
+    SPI2LPEN,
+    SPI2,
+}
+
+ // not stm32_mcu = "stm32f410",
+ #[cfg(any(
+    stm32_mcu = "STM32F401",
```
Comment (rmja) [inline src/periph/spi/f4.rs]: This should probaly be lowercase?

File: src/periph/spi/lib.rs [hunk 2]
```diff
@@ -29,6 +29,21 @@ mod f1;
 ))]
 mod l4_all;
 
+#[cfg(any(
+    stm32_mcu = "STM32F401",
```
Comment (rmja) [inline src/periph/spi/lib.rs]: Again, this should probaly be lowercase?

File: src/periph/spi/lib.rs [hunk 3]
```diff
@@ -51,3 +66,19 @@ pub use self::f1::*;
     stm32_mcu = "stm32l4s9"
 ))]
 pub use self::l4_all::*;
+
+
+#[cfg(any(
+    stm32_mcu = "STM32F401",
```
Comment (rmja) [inline src/periph/spi/lib.rs]: Yet again

**Comment by alfredch** [comment]
@lysseslukker 
I started reviewing your PR. The register and field declarations are sometimes not accurately following the reference manual (RM0368 in the case of the STM32F401) and the SVD. This PR will need many corrections.
How do you want to proceed? You can get closer to an acceptable PR when you run the `just test` command for the stm32f401 (setting `stm32_mcu := 'stm32f401'` in the `Justfile`) and make the necessary corrections until `just test` compiles without errors.

**Comment by lysseslukker** [comment]
alfredch. All test has been passed for all stm32f4. For some reason differs the naming for SPI(4-6)EN(R) for all device except stm32f413, that keeps the same naming as SPI(1-3). I dont know if the solution is the most elegant, but it does the job.

**Comment by alfredch** [comment]
Good job! I will review it tomorrow. Sometimes (unfortunately rather often), the SVD files have errors or omissions. For instance: the stm32f401 has the SPI4 device (which you skipped?), but the downloadable STM32F401.svd is missing SPI4EN (and probably more). In such cases, we patch the SVD file (functions in /drone-stm32-map/svd/src. Sometimes, I patched the SVD file directly because it is so annoying to write all those small patch functions.

**Comment by lysseslukker** [comment]
I skipped it because it was not in the svd file. I will look at it tomorrow.

File: src/periph/spi/lib.rs [hunk 4]
```diff
@@ -29,6 +29,21 @@ mod f1;
 ))]
 mod l4_all;
 
+#[cfg(any(
+    stm32_mcu = "stm32F401",
```
Comment (rmja) [inline src/periph/spi/lib.rs]: This should have a lowercase "f" (repeated multiple places)

**Comment by alfredch** [comment]
@lysseslukker When a field is named differently in the SVD than in the Ref.Man., it's better to use the Ref.Man name. The macro allows aliasing so that it refers to the name in the SVD:
```diff
@@ -59,7 +59,7 @@ periph!{
         }
         SR{
             0x20 RwRegBitBand;
-            TIFRFE {RoRwRegFieldBitBand} //called FRF in ref manual
+            FRF { RoRwRegFieldBitBand }
             BSY { RoRwRegFieldBitBand }
             OVR { RoRwRegFieldBitBand }
             MODF { RoRwRegFieldBitBand }
@@ -160,7 +160,7 @@ macro_rules! map_spi {
                 }
                 SR {
                     SR;
-                    TIFRFE { TIFRFE }
+                    FRF { TIFRFE }
                     BSY { BSY }
                     CHSIDE { CHSIDE }
                     CRCERR { CRCERR }
```

**Comment by alfredch** [comment]
@lysseslukker 
Please format the code consistently. For instance, we use spaces in the macro arguments block, so 
```SPILPEN {RwRwRegFieldBitBand}```
should be written with spaces:
```SPILPEN { RwRwRegFieldBitBand }```
We plan to publish something like https://rust-lang.github.io/api-guidelines/ for Drone-OS, but in the meantime, we use the style guides from Rust library developers.

**Comment by lysseslukker** [comment]
Sorry for the small mistakes, I will correct.

**Review by alfredch — CHANGES_REQUESTED** [review]
I have added the I2S variants for all stm32f4 that we support. I will soon submit a PR against your lysseslukker/drone-stm32-map branch f4-spi

**Comment by rmja** [comment]
For what it is worth. This works fine for me with SPI1 and SPI2 on stm32f429

**Comment by valff** [comment]
Thank you for the good job!

**Pull request merged**